### PR TITLE
doc: add num_workers workflow example

### PR DESCRIPTION
**Pull request by Tonyrj3268**
# Description

Added tutorial on using num_workers in workflow

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: docs/docs/examples/workflow/parallel_execution.ipynb [hunk 1]
```diff
@@ -0,0 +1,212 @@
+{
```
Comment (nerdai) [inline docs/docs/examples/workflow/parallel_execution.ipynb:1]: <p>nit: are the "#" marks supposed to be used for markdown headers?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/15262/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (nerdai) [inline docs/docs/examples/workflow/parallel_execution.ipynb:1]: <p>If easy to add, I think might be beneficial to include a Workflow that doesn't set <code>num_workers</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/15262/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Review by nerdai — COMMENTED** [review]
Thanks @Tonyrj3268  -- left a few minor comments in your notebook.

**Comment by nerdai** [comment]
Looks good to me @Tonyrj3268! Thanks for making the changes, is there anything else that we need to do here or are we good to merge?

**Review by nerdai — APPROVED** [review]

**Comment by Tonyrj3268** [comment]
@nerdai 
Thanks for your review. I believe the current examples are sufficient to express my usage of this topic. Yes, please help me merge.

btw, I noticed that when we don't declare pass_context=True in the @step() decorator, but include ctx in the function parameters (the second step), this error occurs. I think this error message is not sufficient to help me find the source of the error. What are your thoughts on this?
Example:
```
class ParallelWorkflow(Workflow):
    
    @step(pass_context=True)
    async def start(self, ctx: Context, ev: StartEvent) -> ProcessEvent:
        data_list = ["A", "B", "C"]
        ctx.data["num_to_collect"] = len(data_list)
        for item in data_list:
            self.send_event(ProcessEvent(data=item))
        return None

    @step(num_workers=3)
    async def process_data(self,ctx:Context, ev: ProcessEvent) -> ResultEvent:
        # Simulate some time-consuming processing
        await asyncio.sleep(random.randint(1, 2))
        result = f"Processed: {ev.data}"
        print(f"Completed processing: {ev.data}")
        return ResultEvent(result=result)

    @step(pass_context=True)
    async def combine_results(self, ctx: Context, ev: ResultEvent) -> StopEvent | None:
        num_to_collect = ctx.data["num_to_collect"]
        results = ctx.collect_events(ev, [ResultEvent] * num_to_collect)
        if results is None:
            return None

        combined_result = ", ".join([event.result for event in results])
        return StopEvent(result=combined_result)
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-32-21beec1ce376>](https://42smiqvn389-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240808-060219_RC00_660777750#) in <cell line: 17>()
     15 print("Start a parallel workflow with setting num_workers in the step of process_data")
     16 start_time = time.time()
---> 17 result = await parallel_workflow.run()
     18 end_time = time.time()
     19 print(f"Workflow result: {result}")

5 frames
[/usr/lib/python3.10/inspect.py](https://42smiqvn389-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20240808-060219_RC00_660777750#) in _bind(self, args, kwargs, partial)
   3099                             msg = 'missing a required argument: {arg!r}'
   3100                             msg = msg.format(arg=param.name)
-> 3101                             raise TypeError(msg) from None
   3102             else:
   3103                 # We have a positional argument to process

TypeError: missing a required argument: 'ev'
```

**Comment by nerdai** [comment]
@Tonyrj3268 ah thanks for flagging. So, if I understand you correctly, the error message is not really intuitive? Yea it doesn't really suggest that you should pass `context`. Maybe we can say something along the lines of "...Try setting `pass_context` to `True`)"

In any case can you file a separate issue for this? I'm going to merge this one now!

**Pull request merged**